File: 2024/15/15.py
#!/usr/bin/env python3
import sys

# Define direction constants using complex numbers
UP = -1j
DOWN = 1j
LEFT = -1
RIGHT = 1

def parse_input(input_str):
    # Split input into blocks separated by two newlines
    blocks = input_str.split('\n\n')
    
    # Parse the map lines
    lines = blocks[0].split('\n')
    map_dict = {}
    
    for y in range(len(lines)):
        for x in range(len(lines[y])):
            # Map each character to its position using complex numbers
            map_dict[x + y * DOWN] = lines[y][x]
    
    # Parse the steps from the second block
    steps = [UP if ch == '^' else LEFT if ch == '<' else RIGHT if ch == '>' else DOWN for ch in blocks[1].replace('\n', '')]
    return map_dict, steps

def try_to_step(map_dict, robot_pos, direction):
    # Save the original state of the map
    original_map = map_dict.copy()
    
    # Handle different types of characters on the map
    if map_dict[robot_pos] == '.':
        return True  # Empty space, can move
    elif map_dict[robot_pos] in ('O', '@'):
        # Move a box or robot
        if try_to_step(map_dict, robot_pos + direction, direction):
            map_dict[robot_pos + direction] = map_dict[robot_pos]
            map_dict[robot_pos] = '.'
            return True
    elif map_dict[robot_pos] == ']':
        # Handle right wall of a box
        return try_to_step(map_dict, robot_pos + LEFT, direction)
    elif map_dict[robot_pos] == '[':
        # Handle left wall of a box
        if direction == LEFT:
            if try_to_step(map_dict, robot_pos + LEFT, direction):
                map_dict[robot_pos + LEFT] = '['
                map_dict[robot_pos] = ']'
                map_dict[robot_pos + RIGHT] = '.'
                return True
        elif direction == RIGHT:
            if try_to_step(map_dict, robot_pos + 2 * RIGHT, direction):
                map_dict[robot_pos] = '.'
                map_dict[robot_pos + RIGHT] = '['
                map_dict[robot_pos + 2 * RIGHT] = ']'
                return True
        else:
            if (try_to_step(map_dict, robot_pos + direction, direction) and 
                try_to_step(map_dict, robot_pos + RIGHT + direction, direction)):
                map_dict[robot_pos] = '.'
                map_dict[robot_pos + RIGHT] = '.'
                map_dict[robot_pos + direction] = '['
                map_dict[robot_pos + direction + RIGHT] = ']'
                return True
    
    # If move is not possible, revert to original state
    map_dict.clear()
    map_dict.update(original_map)
    return False

def solve(input_str):
    # Parse the input and get the initial state of the map and steps
    map_dict, steps = parse_input(input_str)
    
    # Find the starting position of the robot
    robot = next(pos for pos, char in map_dict.items() if char == '@')
    
    # Process each step
    for direction in steps:
        if try_to_step(map_dict, robot, direction):
            robot += direction
    
    # Calculate the result based on the final positions of boxes and robots
    result = sum(box.real + 100 * box.imag for box in map_dict if map_dict[box] in ('[', 'O'))
    return int(result)

def scale_up(input_str):
    # Scale up the input by replacing characters with larger representations
    return input_str.replace('#', '##').replace('.', '..').replace('O', '[]').replace('@', '@.')

if __name__ == "__main__":
    inp = sys.stdin.read()
    print(solve(inp))  # Solve the original input
    print(solve(scale_up(inp)))  # Solve the scaled-up input

File: 2024/15/15.rs
use std::collections::HashMap;
use std::io::{self, Read};

const UP: (isize, isize) = (0, -1);
const DOWN: (isize, isize) = (0, 1);
const LEFT: (isize, isize) = (-1, 0);
const RIGHT: (isize, isize) = (1, 0);

fn parse_input(input_str: &str) -> (HashMap<(isize, isize), char>, Vec<(isize, isize)>) {
    let blocks: Vec<&str> = input_str.split("\n\n").collect();
    let lines: Vec<&str> = blocks[0].lines().collect();
    let mut map_dict = HashMap::new();

    for (y, line) in lines.iter().enumerate() {
        for (x, c) in line.chars().enumerate() {
            map_dict.insert((x as isize, y as isize), c);
        }
    }

    let steps: Vec<(isize, isize)> = blocks[1]
        .replace('\n', "")
        .chars()
        .map(|ch| match ch {
            '^' => UP,
            '<' => LEFT,
            '>' => RIGHT,
            'v' => DOWN,
            _ => (0, 0),
        })
        .collect();

    (map_dict, steps)
}

fn try_to_step(
    map_dict: &mut HashMap<(isize, isize), char>,
    robot_pos: &(isize, isize),
    direction: &(isize, isize),
) -> bool {
    let original_map = map_dict.clone();
    
    if *map_dict.get(robot_pos).unwrap() == '.' {
        return true;
    } else if *map_dict.get(robot_pos).unwrap() == 'O' || *map_dict.get(robot_pos).unwrap() == '@' {
        let new_pos = (robot_pos.0 + direction.0, robot_pos.1 + direction.1);
        if try_to_step(map_dict, &new_pos, direction) {
            map_dict.insert(new_pos, *map_dict.get(robot_pos).unwrap());
            map_dict.insert(*robot_pos, '.');
            return true;
        }
    } else if *map_dict.get(robot_pos).unwrap() == ']' {
        let new_pos = (robot_pos.0 - 1, robot_pos.1);
        return try_to_step(map_dict, &new_pos, direction);
    } else if *map_dict.get(robot_pos).unwrap() == '[' {
        let dir_x = direction.0;
        let dir_y = direction.1;

        if (dir_x, dir_y) == LEFT {
            let new_pos_1 = (robot_pos.0 - 1, robot_pos.1);
            if try_to_step(map_dict, &new_pos_1, direction) {
                map_dict.insert(new_pos_1, '[');
                map_dict.insert(*robot_pos, ']');
                map_dict.insert((robot_pos.0 + 1, robot_pos.1), '.');
                return true;
            }
        } else if (dir_x, dir_y) == RIGHT {
            let new_pos_2 = (robot_pos.0 + 2, robot_pos.1);
            if try_to_step(map_dict, &new_pos_2, direction) {
                map_dict.insert(*robot_pos, '.');
                map_dict.insert((robot_pos.0 + 1, robot_pos.1), '[');
                map_dict.insert(new_pos_2, ']');
                return true;
            }
        } else {
            let new_pos_1 = (robot_pos.0 + dir_x, robot_pos.1 + dir_y);
            let new_pos_2 = (robot_pos.0 + 1 + dir_x, robot_pos.1 + dir_y);

            if try_to_step(map_dict, &new_pos_1, direction) && try_to_step(map_dict, &new_pos_2, direction) {
                map_dict.insert(*robot_pos, '.');
                map_dict.insert((robot_pos.0 + 1, robot_pos.1), '.');
                map_dict.insert(new_pos_1, '[');
                map_dict.insert(new_pos_2, ']');
                return true;
            }
        }
    }

    *map_dict = original_map.clone();
    false
}

fn solve(input_str: &str) -> isize {
    let (mut map_dict, steps) = parse_input(input_str);
    let mut robot = map_dict.iter().find(|&(_, &c)| c == '@').unwrap().0;

    for direction in &steps {
        if try_to_step(&mut map_dict, &robot, direction) {
            robot.0 += direction.0;
            robot.1 += direction.1;
        }
    }

    let result: isize = map_dict
        .iter()
        .filter(|&(_, &c)| c == '[' || c == 'O')
        .map(|(&(x, y), _)| x + 100 * y)
        .sum();

    result
}

fn scale_up(input_str: &str) -> String {
    input_str
        .replace('#', "##")
        .replace('.', "..")
        .replace('O', "[]")
        .replace('@', "@.")
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();

    println!("{}", solve(&input));
    println!("{}", solve(&scale_up(&input)));
}